app/extractors/IndividualBeneficiaryExtractor.scala

```diff
@@ -82,34 +131,42 @@ class IndividualBeneficiaryExtractor @Inject()() {
 
   private def extractShareOfIncome(individualBeneficiary: IndividualBeneficiary,
                                    answers: UserAnswers) : Try[UserAnswers] = {
-    individualBeneficiary.income match {
-      case Some(income) =>
-        answers.set(IncomeDiscretionYesNoPage, false)
-          .flatMap(_.set(IncomePercentagePage, income.toInt))
-      case None =>
-        // Assumption that user answered yes as the share of income is not provided
-        answers.set(IncomeDiscretionYesNoPage, true)
+    if (answers.isTaxable) {
```

Comment:
There is a BeneficiaryExtractor trait you could extend that has the share of income, address, and country of residence methods defined. You just need to override the page objects. We also have something similar in maintain-a-trust-frontend to refactor the country of residence and country of nationality methods if you want to implement something similar? (https://github.com/hmrc/maintain-a-trust-frontend/blob/7dd2139d9fa5446a479cd1575b126ce66e869fa7/app/mapping/PlaybackExtractor.scala#L67-L108)

Reply:
Updated